active-rfcs/0009-global-api-change.md

```diff
@@ -160,6 +176,21 @@ app.config.isCustomElement = tag => tag.startsWith('ion-')
 
 - This will be a new top-level option in the Vue CLI config.
 
+## Attaching Globally Shared Instance Properties
+
+In 2.x, it was possible to inject globally shared instance properties by simply attaching them to `Vue.prototype`.
+
+In Vue 3, since the global `Vue` is no longer a constructor, this is no longer supported. Instead, shared instance properties should be attached to an app instance's `config.globalProperties` instead:
+
+``` js
+// Before
+Vue.prototype.$http = () => {}
+
+// After
+const app = createApp()
+app.config.globalProperties.$http = () => {}
```

Comment:
how would we access this inside the `setup` function?

Reply:
This is essentially a legacy escape hatch and not for `setup` function. You should just import things directly or use `provide/inject` in setup functions.

Reply:
This won't work for legacy apps that use multiple instances and apply `Vue.use` globally since you'd have to run `app.use` in every one of them. If that's required only for legacy plugins shouldn't that be included only in the compatibility build with global registration on the prototype as usual? That way you won't have to rewrite that much and stay with global `Vue.use`. Why would you need that hack for anything but legacy plugins? (Assuming that the new strategy for plugins is to utilize `provide`\\`inject`).

Reply:
The "legacy" term I used was incorrect - it should be for "options/this-based usage". Even with `setup` function this can still be used to expose globally available properties to templates.

With the global API change there is no once-for-all global configs anymore. 2.x multi-root-instance apps will have to be updated to configure each root instance individually. If all root instances share the same config, then it can be extracted into a shared function.